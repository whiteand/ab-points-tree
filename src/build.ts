import { append } from './append';
import { IBuildConfig } from './types';

export function* build<Point extends number[]>(
  xs: Iterable<number>,
  fn: (x: number) => Point,
  { chunkSize, chunkPredicate }: IBuildConfig<Point> = {},
) {
  let index = -1;
  let currentChunkSize = 0;
  let tree = null;
  for (const x of xs) {
    index++;
    currentChunkSize++;
    const point = fn(x);
    tree = append(tree, point);
    if (
      (chunkSize !== undefined && currentChunkSize >= chunkSize) ||
      (chunkPredicate && chunkPredicate(point, index, tree, currentChunkSize))
    ) {
      currentChunkSize = 0;
      yield tree;
    }
  }
  if (currentChunkSize > 0) {
    yield tree;
  }
}
