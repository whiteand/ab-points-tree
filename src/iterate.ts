import { ITree } from './types';

export function* iterate<Point extends number[]>(
  tree: ITree<Point> | null,
): Generator<Point, void> {
  if (!tree) {
    return;
  }
  yield* iterate(tree.left);
  yield tree.point;
  yield* iterate(tree.right);
}

export function iterateFunc<Point extends number[]>(
  tree: ITree<Point> | null,
  callback: (point: Point) => void,
) {
  if (!tree) return null;

  iterateFunc(tree.left, callback);
  callback(tree.point);
  iterateFunc(tree.right, callback);
}
