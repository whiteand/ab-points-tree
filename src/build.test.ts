import { build } from './build';
import { iterate } from './iterate';

describe('build function', () => {
  test('without config returns full tree', () => {
    const tree = build([3, 2, 4, 1, 5], x => [x, x * x]);
    expect([...iterate(tree.next().value || null)]).toMatchSnapshot();
  });
  test('without config returns tree with each chunk calculated', () => {
    const tree = build([3, 2, 4, 1, 5], x => [x, x * x], { chunkSize: 2 });
    expect([...iterate(tree.next().value || null)]).toMatchSnapshot();
    expect([...iterate(tree.next().value || null)]).toMatchSnapshot();
    expect([...iterate(tree.next().value || null)]).toMatchSnapshot();
    expect([...iterate(tree.next().value || null)]).toMatchSnapshot();
  });
  test('without config returns tree with each chunk calculated from predicate', () => {
    const tree = build([3, 2, 4, 1, 5], x => [x, x * x], {
      chunkPredicate: point => point[0] < 4,
    });
    expect([...iterate(tree.next().value || null)]).toMatchSnapshot();
    expect([...iterate(tree.next().value || null)]).toMatchSnapshot();
    expect([...iterate(tree.next().value || null)]).toMatchSnapshot();
    expect([...iterate(tree.next().value || null)]).toMatchSnapshot();
    expect([...iterate(tree.next().value || null)]).toMatchSnapshot();
  });
});
