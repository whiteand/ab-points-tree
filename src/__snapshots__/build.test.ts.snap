// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build function without config returns full tree 1`] = `
Array [
  Array [
    1,
    1,
  ],
  Array [
    2,
    4,
  ],
  Array [
    3,
    9,
  ],
  Array [
    4,
    16,
  ],
  Array [
    5,
    25,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated 1`] = `
Array [
  Array [
    2,
    4,
  ],
  Array [
    3,
    9,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated 2`] = `
Array [
  Array [
    1,
    1,
  ],
  Array [
    2,
    4,
  ],
  Array [
    3,
    9,
  ],
  Array [
    4,
    16,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated 3`] = `
Array [
  Array [
    1,
    1,
  ],
  Array [
    2,
    4,
  ],
  Array [
    3,
    9,
  ],
  Array [
    4,
    16,
  ],
  Array [
    5,
    25,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated 4`] = `Array []`;

exports[`build function without config returns tree with each chunk calculated from predicate 1`] = `
Array [
  Array [
    3,
    9,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated from predicate 2`] = `
Array [
  Array [
    2,
    4,
  ],
  Array [
    3,
    9,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated from predicate 3`] = `
Array [
  Array [
    1,
    1,
  ],
  Array [
    2,
    4,
  ],
  Array [
    3,
    9,
  ],
  Array [
    4,
    16,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated from predicate 4`] = `
Array [
  Array [
    1,
    1,
  ],
  Array [
    2,
    4,
  ],
  Array [
    3,
    9,
  ],
  Array [
    4,
    16,
  ],
  Array [
    5,
    25,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated from predicate 5`] = `Array []`;

exports[`build function without config returns tree with each chunk calculated from predicate 6`] = `
Array [
  Array [
    3,
    9,
  ],
  Array [
    10,
    100,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated from predicate 7`] = `
Array [
  Array [
    2,
    4,
  ],
  Array [
    3,
    9,
  ],
  Array [
    10,
    100,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated from predicate 8`] = `
Array [
  Array [
    1,
    1,
  ],
  Array [
    2,
    4,
  ],
  Array [
    3,
    9,
  ],
  Array [
    4,
    16,
  ],
  Array [
    10,
    100,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated from predicate 9`] = `
Array [
  Array [
    1,
    1,
  ],
  Array [
    2,
    4,
  ],
  Array [
    3,
    9,
  ],
  Array [
    4,
    16,
  ],
  Array [
    5,
    25,
  ],
  Array [
    10,
    100,
  ],
]
`;

exports[`build function without config returns tree with each chunk calculated from predicate 10`] = `Array []`;
